File: backend/src/Gorodki.Domain/Leagues/LeagueRules.cs
namespace Gorodki.Domain.Leagues;

/// <summary>Предел средней скорости за окно: если за последние <c>WindowSeconds</c> средняя скорость выше — это не бег (не велосипед).</summary>
public sealed record SpeedLimit(double WindowSeconds, double MaxKilometersPerHour);

/// <summary>«Разгон машины»: скорость выросла на <c>DeltaKilometersPerHour</c> за <c>WithinSeconds</c> и дошла до <c>ReachingKilometersPerHour</c>.</summary>
public sealed record CarLaunchRule(double DeltaKilometersPerHour, double WithinSeconds, double ReachingKilometersPerHour);

/// <summary>«Транспорт по датчикам»: доля времени automotive за окно и одновременно высокая скорость.</summary>
public sealed record VehicleShareRule(double WindowSeconds, double MinShare, double SpeedWindowSeconds, double MinKilometersPerHour);

/// <summary>
/// Пороги античита одной лиги (PLAN.md, §3.9). Зеркало <c>LeagueRules</c> из GameCore (Swift): те же имена и та же форма JSON,
/// чтобы телефон читал этот раздел конфига напрямую. Совпадение проверяют тесты обеих сторон по <c>contracts/game-config.v1.json</c>.
/// </summary>
public sealed record LeagueRules
{
    // Слой 1 — сама точка.

    /// <summary>Хуже этой точности точка для захвата не используется (для первой петли новичка — отдельный порог).</summary>
    public double MaxAccuracyMeters { get; init; } = 25;

    /// <summary>Точка старше этого — устарела.</summary>
    public double MaxFixAgeSeconds { get; init; } = 10;

    /// <summary>Скачок быстрее этого — «телепорт», след рвётся (180 км/ч, калибруется на полевом тесте).</summary>
    public double TeleportMetersPerSecond { get; init; } = 50;

    // Слой 2 — отрезки.

    public IReadOnlyList<SpeedLimit> SpeedLimits { get; init; } = [];

    /// <summary>Бег: «транспорт» по датчикам дольше этого — разрыв.</summary>
    public double? VehicleSeconds { get; init; }

    /// <summary>Бег: «велосипед» по датчикам дольше этого — разрыв и предложение перейти в лигу «Вело».</summary>
    public double? CyclingSeconds { get; init; }

    public VehicleShareRule? VehicleShare { get; init; }

    public CarLaunchRule? CarLaunch { get; init; }

    /// <summary>Бег: ноль шагов за это окно при скорости не ниже <see cref="NoStepsMinSpeed"/> — разрыв.</summary>
    public double? NoStepsWindowSeconds { get; init; }

    public double NoStepsMinSpeed { get; init; } = 1.0;

    /// <summary>Бег: допустимая длина шага <c>[от, до]</c>, метры (так Swift кодирует <c>ClosedRange</c>).</summary>
    public IReadOnlyList<double>? StrideMeters { get; init; }

    /// <summary>Окно, за которое считается длина шага.</summary>
    public double StrideWindowSeconds { get; init; } = 30;

    /// <summary>Лига «Бег»: ходьба и бег.</summary>
    public static LeagueRules Run { get; } = new()
    {
        SpeedLimits = [new(30, 25), new(300, 19)],
        VehicleSeconds = 20,
        CyclingSeconds = 30,
        NoStepsWindowSeconds = 20,
        StrideMeters = [0.3, 2.2],
    };

    /// <summary>Лига «Вело».</summary>
    public static LeagueRules Bike { get; } = new()
    {
        SpeedLimits = [new(5, 60), new(30, 48), new(60, 42), new(300, 36)],
        VehicleShare = new(WindowSeconds: 120, MinShare: 0.6, SpeedWindowSeconds: 60, MinKilometersPerHour: 25),
        CarLaunch = new(DeltaKilometersPerHour: 25, WithinSeconds: 6, ReachingKilometersPerHour: 35),
    };
}

File: ios/Packages/GameCore/Sources/GameCore/Tracking/LeagueRules.swift
/// Лига: у бега и велосипеда свои правила и своя карта (PLAN.md, D16).
public enum League: String, Codable, Sendable, CaseIterable {
    case run
    case bike
}

/// Предел средней скорости за окно: если за последние `windowSeconds` средняя скорость выше — это не бег (не велосипед).
public struct SpeedLimit: Hashable, Codable, Sendable {
    public var windowSeconds: Double
    public var maxKilometersPerHour: Double

    public init(windowSeconds: Double, maxKilometersPerHour: Double) {
        self.windowSeconds = windowSeconds
        self.maxKilometersPerHour = maxKilometersPerHour
    }
}

/// «Разгон машины»: скорость выросла на `deltaKilometersPerHour` за `withinSeconds` и дошла до `reachingKilometersPerHour`.
public struct CarLaunchRule: Hashable, Codable, Sendable {
    public var deltaKilometersPerHour: Double
    public var withinSeconds: Double
    public var reachingKilometersPerHour: Double
}

/// «Транспорт по датчикам»: доля времени automotive за окно и одновременно высокая скорость.
public struct VehicleShareRule: Hashable, Codable, Sendable {
    public var windowSeconds: Double
    public var minShare: Double
    public var speedWindowSeconds: Double
    public var minKilometersPerHour: Double
}

/// Пороги античита одной лиги (PLAN.md, §3.9). Хранятся в игровом конфиге, здесь — значения по умолчанию.
public struct LeagueRules: Hashable, Codable, Sendable {
    // Слой 1 — сама точка.
    /// Хуже этой точности точка для захвата не используется (для первой петли новичка — 35 м).
    public var maxAccuracyMeters: Double = 25
    /// Точка старше этого — устарела.
    public var maxFixAgeSeconds: Double = 10
    /// Скачок быстрее этого — «телепорт», след рвётся. Начальное значение (180 км/ч), калибруется на полевом тесте.
    public var teleportMetersPerSecond: Double = 50

    // Слой 2 — отрезки.
    public var speedLimits: [SpeedLimit]
    /// Бег: «транспорт» по датчикам дольше этого — разрыв.
    public var vehicleSeconds: Double?
    /// Бег: «велосипед» по датчикам дольше этого — разрыв и предложение перейти в лигу «Вело».
    public var cyclingSeconds: Double?
    public var vehicleShare: VehicleShareRule?
    public var carLaunch: CarLaunchRule?
    /// Бег: ноль шагов за это окно при скорости не ниже `noStepsMinSpeed` — разрыв.
    public var noStepsWindowSeconds: Double?
    public var noStepsMinSpeed: Double = 1.0
    /// Бег: длина шага вне этих пределов — разрыв.
    public var strideMeters: ClosedRange<Double>?
    /// Окно, за которое считается длина шага.
    public var strideWindowSeconds: Double = 30

    /// Лига «Бег»: ходьба и бег.
    public static let run = LeagueRules(
        speedLimits: [
            SpeedLimit(windowSeconds: 30, maxKilometersPerHour: 25),
            SpeedLimit(windowSeconds: 300, maxKilometersPerHour: 19),
        ],
        vehicleSeconds: 20,
        cyclingSeconds: 30,
        vehicleShare: nil,
        carLaunch: nil,
        noStepsWindowSeconds: 20,
        strideMeters: 0.3...2.2
    )

    /// Лига «Вело».
    public static let bike = LeagueRules(
        speedLimits: [
            SpeedLimit(windowSeconds: 5, maxKilometersPerHour: 60),
            SpeedLimit(windowSeconds: 30, maxKilometersPerHour: 48),
            SpeedLimit(windowSeconds: 60, maxKilometersPerHour: 42),
            SpeedLimit(windowSeconds: 300, maxKilometersPerHour: 36),
        ],
        vehicleSeconds: nil,
        cyclingSeconds: nil,
        vehicleShare: VehicleShareRule(
            windowSeconds: 120, minShare: 0.6, speedWindowSeconds: 60, minKilometersPerHour: 25),
        carLaunch: CarLaunchRule(deltaKilometersPerHour: 25, withinSeconds: 6, reachingKilometersPerHour: 35),
        noStepsWindowSeconds: nil,
        strideMeters: nil
    )

    public static func `default`(for league: League) -> LeagueRules {
        switch league {
        case .run: .run
        case .bike: .bike
        }
    }
}
